improver/psychrometric_calculations/cloud_condensation_level.py

```diff
@@ -0,0 +1,130 @@
+    level from near-surface values of temperature, pressure and humidity mixing
+    ratio.
+    """
+
+    def __init__(self, model_id_attr: str = None):
+        """
+        Set up class
+
+        Args:
+            model_id_attr:
+                Name of model ID attribute to be copied from source cubes to output cube
+        """
+        self.model_id_attr = model_id_attr
+        self.temperature, self.pressure, self.humidity = None, None, None
+
+    def _make_ccl_cube(self, data: np.ndarray, pressure: np.ndarray) -> Cube:
+        """Puts the data array into a CF-compliant cube"""
+        attributes = {}
+        if self.model_id_attr:
+            attributes[self.model_id_attr] = self.temperature.attributes[
+                self.model_id_attr
+            ]
+        cube = create_new_diagnostic_cube(
+            "air_temperature_at_condensation_level",
+            "K",
+            self.temperature,
+            mandatory_attributes=generate_mandatory_attributes(
+                [self.temperature, self.pressure, self.humidity]
+            ),
+            optional_attributes=attributes,
+            data=data,
+        )
+        pressure_coord = AuxCoord(
+            standard_name="air_pressure", units="Pa", points=pressure
+        )
+        spatial_dims = range(self.temperature.ndim)
+        cube.add_aux_coord(pressure_coord, spatial_dims)
+        return cube
+
+    def _iterate_to_ccl(self) -> Tuple[np.ndarray, np.ndarray]:
+        """Uses a Newton iterator to find the pressure level where the
+        adiabatically-adjusted temperature equals the saturation temperature.
+        Returns pressure and temperature arrays."""
+
+        def humidity_delta(p2, p, t, q):
+            """For a target pressure guess, p2, and origin p, t and q, return the
+            difference between q and q_sat(t2, p2)"""
+            t2 = dry_adiabatic_temperature(t, p, p2)
+            return q - saturated_humidity(t2, p2)
+
+        ccl_pressure = newton(
+            humidity_delta,
+            self.pressure.data.copy(),
+            args=(self.pressure.data, self.temperature.data, self.humidity.data),
+            tol=100,
+            maxiter=20,
+        ).astype(np.float32)
+        ccl_temperature = dry_adiabatic_temperature(
+            self.temperature.data, self.pressure.data, ccl_pressure
+        )
+        return ccl_pressure, ccl_temperature
+
+    def process(self, cubes: List[Cube]) -> Cube:
+        """
+        Calculates the cloud condensation level from the near-surface inputs.
+
+        Args:
+            cubes:
+                Cubes, in this order, of temperature (K), pressure (Pa)
+                and humidity mixing ratio (kg kg-1)
+
+        Returns:
+            Cube of cloud condensation level
+
+        """
+        self.temperature, self.pressure, self.humidity = cubes
+        ccl_pressure, ccl_temperature = self._iterate_to_ccl()
+        return self._make_ccl_cube(ccl_temperature, ccl_pressure)
```

Comment:
The only reason is to return a single cube. It is reasonable to contain a pressure coordinate though, we have them elsewhere for `temperature_on_pressure_levels` for example. I have checked it with Bruce and he is happy.

Reply:
If we choose to output two cubes here, instead of one, then my recommended approach would be:

- Refactor the `CloudCondensationLevel._make_ccl_cube` so that it takes in a `data` array and `is_temperature: bool`, but not `pressure`. Remove the lines that define and add the `pressure_coord` to the cube and add an `if` near the beginning to switch between the name and units for temperature or pressure depending on the value of `is_temperature`.
- Modify `process` to call this method twice, once for each array and return both cubes as a tuple. The CLI should not need changing, as I can see examples of plugins returning multiple cubes and these being written to one file, e.g. `EstimateCoefficientsForEnsembleCalibration`.
- Update unit tests and acceptance tests appropriately.
- Update the cloud top temperature PR. This currently loads two files and the CLI separates them based on the cube names. If one of the files contains two cubes, a different approach will be required. I would like that approach to continue to be agnostic of the order the input files are specified, but I don't know if there is a ready-made solution for this, I suspect not, looking at the `nowcast-accumulate` CLI.
- The cloud top temperature PR plugin will also need updating to expect an extra cube, and to get the pressure data array from it instead of from a coord.

Reply:
Responding to @bayliffe comment "I ... would be interested in @bjwheltor's reasoning for going with this approach". I din't actually go with this approach, @MoseleyS did, and I accepted it, if it met 2 criteria:

1. It is being used as a coordinate value (to directly support the interpretation on the main variable), and not just as a way of putting 2 variables in one cube/file.

2. It won't be passed to users.

A CF auxiliary coordinate is define as any netCDF variable that contains coordinate data, but is not a coordinate variable, which doesn't help that much as it brings it down to my point 1 (above). On this point, Stephen responded "The auxilliary coordinate is being used as a way of describing coincident values of T and P. There's also an equivalent altitude that isn't being generated here", which on re-reading makes me slightly unsure (although I think we also discussed the usage in person at the time). As @bayliffe says, the inclusion of orography in file with surface temperature is almost an exact analogue for this (apart form the fact that the orography does not change each timestep) - and some people would put orography in the same file as the primary variable, but more as supporting data than a coordinate variable.

Stephen confirmed the files will not be passed to users, but reading @bayliffe comments, I think I should have included a third requirement that:

4. it won't complicate or restrict the handling of cubes or files through IMPROVER processing or visualisation steps.

On reflection, I think if fails criterium 3 (as it potentially complicates other processing), and probably fails criterium 1, as, in @MoseleyS own words it is "a way of describing coincident values of T and P", which suggests these variables are on equal footing, rather than one just being coordinate variable.

Part of my reason for not challenging the proposal too hard, was my awareness of "the convention to return a single cube" - so it felt like a trade-off between doing something a bit specialised in place or non-standard in another - and the amount of additional work it might involve to separate out pressure. However, give the further discussion on this (and following on from my comments above), I would be more comfortable is we did split temperature and pressure into two separate cubes.

Reply:
Ok. I will make these changes.